### DISCUSSION TITLE
Add setting to disable showing hidden/private modules in import statement completion list for 3rd party libraries

### DISCUSSION THREAD

**Discussion by laundmo** [microsoft/pylance-release#4518] [Enhancement]
<h3> Environment data </h3>

-   Language Server version: 2023.6.21
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: tried both basic and strict

<h3> Code Snippet </h3>

```python
from mytypedlib.
```

<h3> Repro Steps </h3>

1. create a typed library called `mytypedlib`
2. make sure it includes a public module, and a private one denoted by the filename starting with `_`
3. ensure it contains and is packaged with a `py.typed`file
4. install it with `pip install .`
5. verify it still includes the `py.typed` file after install (checking in site-packages)
6. create a new python file
7. start importing from `mytypedlib` by starting to type `from mytypedlib.`
8. trigger completions

<h3> Expected behavior </h3>

The completions should not include modules prefixed with `_` should not show up, as per: https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface they should be considered private.

<h3> Actual behavior </h3>

The autocompletion shows modules prefixed with `_`

<h3> Logs </h3>

i have not included trace logs as it would leak private information. I can provide them with some effort if others can't reproduce this.

**Comment by rchiodo** [comment]
Thanks for the issue but I can't seem to reproduce the problem.

I have a set of files like so:

![image](https://github.com/microsoft/pylance-release/assets/19672699/81882fc9-8af4-4668-919e-e2000a59e561)

In the `__init__.py` I exported the public class from `test.py`.

When I go to autocomplete the library I only see the stuff I exported from init:

![image](https://github.com/microsoft/pylance-release/assets/19672699/f3b7d201-27bc-4a54-ac24-6e82414f807f)

**Comment by rchiodo** [comment]
If the `_privatetest.py` was included, it would have shown this:

![image](https://github.com/microsoft/pylance-release/assets/19672699/b66f5e58-841a-4f3f-a52e-c90776f8eaea)

That's controlled by what's exported from the `__init__.py` though. So maybe your `__init__.py` is reexporting the private modules?

**Comment by laundmo** [comment]
you're looking at the wrong autocomplete: the issue is specific to the 

`from mylib.|` autocomplete, where your cursor is directly after the dot, where i have written a pipe character.

**Comment by rchiodo** [comment]
Ah you mean this:

![image](https://github.com/microsoft/pylance-release/assets/19672699/0cefe46a-8a92-49c1-8451-059b38a0b9f6)

Seems like a bug. I'm guessing import module completions don't exclude privates on purpose though as they are valid to be imported.

Moving to triage to discuss.

**Comment by erictraut** [comment]
This might be related to [this bug](https://github.com/microsoft/pylance-release/issues/4059).

**Comment by debonte** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

Maybe we should add a setting to control this behavior -- should `_` prefixed symbols be shown in the completion list.
Also, maybe the `_`/`__` prefixed symbols should be sorted to the end of the list.

**Reply by erictraut** [reply]
If the symbol is imported from a `py.typed` library and the author of the library has taken the time to indicate which symbols are private and which are public, it seems to me that pylance should honor that information and omit completion suggestions that violate the interface contract of the library. It's different when dealing with a non-`py.typed` library or user code, but this issue is specifically about `py.typed` libraries.

**Comment by laundmo** [comment]
I don't agree with making this a discussion. it's documented in Pyright how it should work

https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface

especially since the behaviour is already correct in many places, just the specific `from` completion isnt.

**Comment by rchiodo** [comment]
The pyright documentation describes what is public and what is private but mentions nothing about how import completions should work. We're worried about the case where somebody is writing a library with py.typed and they want to import private modules. 

I think at a minimum we'll make the private modules sort at the bottom.

**Reply by erictraut** [reply]
If a user of a library wants to violate the public interface contract of a library they're consuming, that's their prerogative, but I don't think tooling should help them do so. IMO, tooling should steer developers away from this practice because it results in code that is not robust (since it can break when internal implementation details change in a library update). It also harms library maintainers in that it prevents them from making internal changes to the library without breaking their users.

**Reply by rchiodo** [reply]
If we could detect that a py.typed library is not being written, but rather just being used, then yes I agree. Users of py.typed libraries should not import private modules.

But if I'm writing a py.typed package, I still need to import private modules from my own package. Can we detect this scenario?

**Reply by erictraut** [reply]
Pyright already has all the logic in place to properly report the use of private modules and symbols under the correct circumstances. It needs to do this to properly handle the `reportPrivateUsage` diagnostic check. The completion provider in pylance is not honoring this information currently.

**Comment by rchiodo** [comment]
Sounds good. Created an issue again.

**Comment by rchiodo** [comment]
A fix for this was submitted to Pyright. 
https://github.com/microsoft/pyright/commit/d2c67d1a56f780b66a9dbd9a4521121d39073892

This should show up in the next prerelease version of Pylance.

**Comment by tjdcs** [comment]
Is this issue alive again, or am I doing something wrong... please advise... 

<img width="649" alt="Screenshot 2025-07-09 at 11 47 42 AM" src="https://github.com/user-attachments/assets/d6e503ff-0702-445a-ad62-fc03286f80dc" />


In this image, I want `__device_implementations` (or similar with a single underscore) at the end of the list. 

Additionally, when someone requests the `CRColorimeter` object, import completion should prefer the re-export which is under `specio.colorimeters` 

If this architecturally sucks... please advise!

**Reply by rchiodo** [reply]
I don't think we can control the sort order, so that was never actually possible. I believe the fix here was to allow private completions as long as the location where they're coming from doesn't have a py.typed file.

**Reply by rchiodo** [reply]
Oh actually I think we can control the [sort order](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem), we just never special cased this situation.

**Reply by tjdcs** [reply]
I get similar behavior when, for example, using auto-import suggestions on the symbol `CRSpectrometer` which I want the ide to suggest the "public" api and use the alias inside of `specio.spectrometers` rather than the `_device_implementations` the reason being... for some purposes it's useful to organize some of the "CR" device implementations in the same module. They share a lot of configuration functions and symbols as they are made by the same manufacturer. But the nature of those devices is different "colorimeters" vs "spectrometers" and I'm trying to organize the devices in my library by those designations...

The code is organized the way I want, but the suggestions are pushing the users to access the device implementations directly leading to confusing code...

But then... aliases themselves are confusing. Maybe I need to go about this a different way.

**Reply by tjdcs** [reply]
Seems I got the correct behavior by changing "include aliases in user files" 

This might look different for users who actually installed via pip rather than are maintainers / using editable dependencies.

**Reply by tjdcs** [reply]
https://github.com/user-attachments/assets/f226cf06-df7b-4b6a-8f97-6047150d7e82

**Reply by tjdcs** [reply]
forgive my `import time` not being in the top of the file hahah.... this is an example script 😅

**Comment by tjdcs** [comment]
@rchiodo, To clarify... one of the things I wanted was affected by that setting and is behaving as expected now. But in the screenshot above where modules are listed during `from specio.|" typing, the `_` entries should be listed last. Seems like that's an outstanding bug / FR?

**Reply by rchiodo** [reply]
I believe that would be a new feature request. We didn't do custom sorting for private entries, so they just sort alphabetically AFAIK.

**Reply by tjdcs** [reply]
https://github.com/user-attachments/assets/0a5f52a0-e661-4a8b-9e19-23f6a7fa893d

Video example.

Thanks! I'll figure out how to open an issue and request it then.
